src/search_span.rs

```diff
@@ -0,0 +1,94 @@
+use std::cmp::Ordering;
+use std::ops::Range;
+use span;
+
+pub type Span = span::Span<span::ZeroIndexed>;
+
+#[derive(Debug)]
+enum SpanKind {
+    InTree, // Span keys tracked inside the BTreeMap
+    Start,  // Start range to search for
+    End     // End range to search for
+}
+
+#[derive(Debug)]
+pub struct SearchSpan(Span, SpanKind);
+
+impl SearchSpan {
+    pub fn range(span: &Span) -> Range<SearchSpan> {
+        SearchSpan(span.clone(), SpanKind::Start)..SearchSpan(span.clone(), SpanKind::End)
+    }
+
+    pub fn in_tree(span: &Span) -> Self {
+        SearchSpan(span.clone(), SpanKind::InTree)
+    }
+
+    pub fn span(&self) -> &Span {
+        &self.0
+    }
+}
+
+impl PartialEq<SearchSpan> for SearchSpan {
+    fn eq(&self, other: &SearchSpan) -> bool {
+        match (&self.1, &other.1) {
+            // In order to be stored inside the BTreeMap, InTree spans should always
+            // equate and compare normally. When we compare start and end search spans,
+            // they should also equate normally to satisfy some properties of the BTreeMap.
+            // All other cases should not equate so we can properly collect the range.
+            // Additional details are in the Ord::cmp implementation.
+            (&SpanKind::InTree, &SpanKind::InTree)
+            | (&SpanKind::Start, &SpanKind::End)
+            | (&SpanKind::End, &SpanKind::Start) => self.0 == other.0,
+            _ => false,
+        }
+
+    }
+}
+impl Eq for SearchSpan { }
```

Comment:
No because that would also require deriving `PartialEq` and `Eq` for `SpanKind`, and I think that would imply the `SpanKind` value would be used to implement the derived `Eq`? See my response to the `Ord` derive.